### PR TITLE
llama : cache llama_token_to_piece

### LINKED ISSUE #7554: Bug: sample time becomes very long when using Llama-3
### What happened?

I was running Llama-3 on 3090 and I encountered the same performance problem in [#1376](https://github.com/abetlen/llama-cpp-python/discussions/1376). 
When using grammar files, sample time becomes very long and GPU utilization dropped from 70%+(when not using grammar) to 10%.
I tried two different fine-tuned version of Llama-3 and the problem remains. 
With Llama-2  there is no such problem. So I believe it is due to some kind of bug in llama.cpp
I offloaded all layers to GPU and I believe I have llama.cpp properly configured.

### Name and Version

version: 2998 (9588f196)
built with cc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0 for x86_64-linux-gnu

### What operating system are you seeing the problem on?

Linux

### Relevant log output

```shell
Llama-3-8B-Instruct with grammar:
llama_print_timings:        load time =     195.81 ms
llama_print_timings:      sample time =    7656.05 ms /    90 runs   (   85.07 ms per token,    11.76 tokens per second)
llama_print_timings: prompt eval time =     192.27 ms /   410 tokens (    0.47 ms per token,  2132.44 tokens per second)
llama_print_timings:        eval time =     944.78 ms /    89 runs   (   10.62 ms per token,    94.20 tokens per second)
llama_print_timings:       total time =    9298.97 ms /   499 tokens

Llama3-8B-Instruct without grammar:
llama_print_timings:        load time =     193.30 ms
llama_print_timings:      sample time =     387.66 ms /   233 runs   (    1.66 ms per token,   601.04 tokens per second)
llama_print_timings: prompt eval time =     192.93 ms /   410 tokens (    0.47 ms per token,  2125.09 tokens per second)
llama_print_timings:        eval time =    2355.86 ms /   232 runs   (   10.15 ms per token,    98.48 tokens per second)
llama_print_timings:       total time =    3277.20 ms /   642 tokens

Llama-2-8B with grammar:
llama_print_timings:        load time =     210.30 ms
llama_print_timings:      sample time =     354.68 ms /    54 runs   (    6.57 ms per token,   152.25 tokens per second)
llama_print_timings: prompt eval time =     209.69 ms /   464 tokens (    0.45 ms per token,  2212.84 tokens per second)
llama_print_timings:        eval time =     492.42 ms /    53 runs   (    9.29 ms per token,   107.63 tokens per second)
llama_print_timings:       total time =    1128.22 ms /   517 tokens

Llama-2-8B without grammar:
llama_print_timings:        load time =     194.85 ms
llama_print_timings:      sample time =     153.25 ms /   367 runs   (    0.42 ms per token,  2394.76 tokens per second)
llama_print_timings: prompt eval time =     194.44 ms /   464 tokens (    0.42 ms per token,  2386.38 tokens per second)
llama_print_timings:        eval time =    3512.26 ms /   366 runs   (    9.60 ms per token,   104.21 tokens per second)
llama_print_timings:       total time =    4094.80 ms /   830 tokens
```

### PR DESCRIPTION
**Pull request by ggerganov**
ref #4218
fix #7554

Build `llama_token_to_piece` caches to speed-up sampling with grammar

### CODE REVIEW

**Comment by skoulik** [comment]
https://github.com/ggerganov/llama.cpp/issues/7554#issuecomment-2134962774

File: llama.cpp [hunk 1]
```diff
@@ -2162,7 +2163,11 @@ struct llama_vocab {
     std::unordered_map<token, id> token_to_id;
     std::vector<token_data>       id_to_token;
 
-    std::vector<id> special_tokens_cache;
+    bool has_cache = false;
+
+    std::vector<id> cache_special_tokens;
```
Comment (HanClinto) [inline llama.cpp:2168]: I like this name change from `special_tokens_cache` to `cache_special_tokens`. 👍

File: llama.cpp [hunk 2]
```diff
@@ -2162,7 +2163,11 @@ struct llama_vocab {
     std::unordered_map<token, id> token_to_id;
     std::vector<token_data>       id_to_token;
 
-    std::vector<id> special_tokens_cache;
+    bool has_cache = false;
```
Comment (HanClinto) [inline llama.cpp:2166]: Is there a mechanism by which the vocab can be loaded _without_ having a cache in place? If not, I'm wondering why `has_cache` is useful right now...?

File: llama.cpp [hunk 3]
```diff
@@ -14408,8 +14421,8 @@ void llama_sample_grammar(struct llama_context * ctx, llama_token_data_array * c
     candidates_grammar.reserve(candidates->size);
 
     for (size_t i = 0; i < candidates->size; ++i) {
-        const llama_token id    = candidates->data[i].id;
-        const std::string piece = llama_token_to_piece(ctx, id, false);
+        const llama_token id      = candidates->data[i].id;
+        const std::string & piece = ctx->model.vocab.cache_token_to_piece.at(id);
```
Comment (HanClinto) [inline llama.cpp:14425]: This is not checking for the presence of a cache with `ctx->model.vocab.has_cache` -- should it?

File: llama.cpp (hunk 2 above)
Comment (ggerganov) [reply llama.cpp]: There was a way to exit early before creating the cache if the tokenizer was unknown. I've removed this path by throwing an exception: [1494a18](https://github.com/ggerganov/llama.cpp/pull/7587/commits/1494a1841ef999cf3a6c4aa5a30ffd2d689a8b79)

There is another path where the GGUF explicitly does not contain a vocabulary: `"no_vocab"`. In that case calling any of the functions that rely on a cache would throw exception due to accessing the caches via `cache.at()`. I think this makes sense

Removed `has_cache` and replaced the unordered maps with vectors

File: llama.cpp [hunk 4]
```diff
@@ -4644,6 +4636,8 @@ static void llm_load_vocab(
             vocab.special_pad_id  = -1;
             vocab.special_cls_id  = -1;
             vocab.special_mask_id = -1;
+        } else {
+            throw std::runtime_error(format("unknown tokenizer: '%s'", tokenizer_model.c_str()));
```
Comment (HanClinto) [inline llama.cpp:4640]: This seems like a good change. A default tokenizer feels like a bad idea, and throwing an exception to require users specify an explicit tokenizer feels like the right choice. 👍

File: llama.cpp [hunk 5]
```diff
@@ -4852,12 +4846,18 @@ static void llm_load_vocab(
...
+       std::vector<llama_vocab::token> cache_token_to_piece_special(n_vocab);
 
-    vocab.has_cache = true;
+       for (uint32_t id = 0; id < n_vocab; ++id) {
+           cache_token_to_piece[id]         = llama_token_to_piece(&model, id, false);
+           cache_token_to_piece_special[id] = llama_token_to_piece(&model, id, true);
+       }
+
+       std::swap(vocab.cache_token_to_piece,         cache_token_to_piece);
+       std::swap(vocab.cache_token_to_piece_special, cache_token_to_piece_special);
```
Comment (HanClinto) [inline llama.cpp:4859]: This feels good. This prevents us from getting into sticky situations if multiple threads are both trying to mess with a half-built vocab. 

As long as the cache isn't empty, then we can be pretty sure that our token is going to be present, and don't need to worry about cache misses.

There are still a few other places where we assume that the cache is going to be present, but in general this feels good I think? I need to look better and understand the vectors by which a vocab could potentially be initialized without a cache.

File: llama.cpp [hunk 6]
```diff
@@ -18305,14 +18306,18 @@ static std::string llama_decode_text(const std::string & text) {
...
 int32_t llama_token_to_piece(const struct llama_model * model, llama_token token, char * buf, int32_t length, bool special) {
-    if (model->vocab.has_cache) {
+    // if we have a cache - use it
+    {
         const auto & cache = special ? model->vocab.cache_token_to_piece_special : model->vocab.cache_token_to_piece;
-        const auto & res = cache.at(token);
-        if (length < (int) res.size()) {
-            return -(int) res.size();
+
+        if (!cache.empty()) {
```
Comment (HanClinto) [inline llama.cpp:18313]: 👍 Nice and clean -- I like it.

File: llama.cpp (hunk 2 above)
Comment (HanClinto) [reply llama.cpp]: Perfect, thank you!

I like all your changes here -- this all feels really good. The only other thing that I'll note is the caveat that I noted on @ochafik 's similar PR in #6811 : 

https://github.com/ggerganov/llama.cpp/pull/6811#discussion_r1589380140

> > I think it'd be simpler to leave it as is and keep it as an area where to potentially squeeze a couple of MB when times are scarce. wdyt?

> This also sounds not unreasonable, but I don't know how to weigh such things. I know I really like grammar-constrained sampling, but I don't know how popular the feature is overall, and is it worth negatively impacting hyper-resource-constrained usages (such as Raspberry Pis or whatnot) vs. grammars? That's what I'm unable to weigh -- I feel like that's a strategic decision that's a bit above my level.

In short, we don't need the cache for situations that don't use grammars, and we're adding a bit of memory usage (n_vocab*2) to every context that we're creating. On most systems this isn't a problem, but on highly-constrained systems (such as Raspberry Pi and whatnot) then this is wasted memory.

How do we weigh the interests of memory-constrained users vs. grammar-enabled users? That's something that I'm not able to make, but overall I think that improving speed performance on grammar-enabled sampling is going to benefit the largest number of people, and the ultra-constrained users are going to be pretty small. We might want to make a note somewhere in a comment that if one is looking for a way to decrease memory usage that they could disable the caching, but beyond that we're probably fine with kicking that can down the road.

**Comment by HanClinto** [comment]
Additional ref: #6811

File: llama.cpp (hunk 3 above)
Comment (HanClinto) [reply llama.cpp:14426]: Comment at https://github.com/ggerganov/llama.cpp/pull/7587#discussion_r1619293050 resolved this question.

File: llama.cpp (hunk 5 above)
Comment (HanClinto) [reply llama.cpp:4859]: Question resolved by comment at https://github.com/ggerganov/llama.cpp/pull/7587#discussion_r1619293050

File: llama.cpp (hunk 2 above)
Comment (ggerganov) [reply llama.cpp]: I added a log for the memory usage of the "token to piece" caches:

```bash
# llama 3
llm_load_vocab: token to piece cache size = 1.5928 MB
```

I think this is completely fine and no need to worry about it for now

Comment (HanClinto) [reply llama.cpp]: Excellent, thank you! That was the one reservation that held me back from fully approving #6811 (I felt that choice required someone with a larger project scope than I have), so I'm very happy to have you weigh in on that.

**Review by HanClinto — APPROVED** [review]
@ochafik may have some insights from his very similar PR, but overall this looks good to me!

**Comment by HanClinto** [comment]
Part of this may be touched in future server refactorings as well.

It would be nice to flag this and not build the cache if there isn't a grammar. However, the server complicates this -- at the time of server initialization, it's not yet known if the user's request is going to include a grammar or not. 

So we need to take the time and memory to initialize the cache no matter what.

All that to say, there is room for additional improvement on this, but ultimately it feels like this is enough of a net win (and the grammar feature is so powerful) that we should include token caching in all cases for now.

**Comment by slaren** [comment]
Can't you initialize the cache the first time `llama_sample_grammar` is called?

**Comment by ggerganov** [comment]
We can lazy initialize the cache, but it's not just the grammar that benefits from it. All calls to `llama_token_to_piece` would be faster too if the cache is pre-computed at the start

File: llama.cpp [hunk 7]
```diff
@@ -18292,69 +18313,83 @@ static std::string llama_decode_text(const std::string & text) {
 
 // does not write null-terminator to buf
 int32_t llama_token_to_piece(const struct llama_model * model, llama_token token, char * buf, int32_t length, bool special) {
+    // if we have a cache - use it
+    {
+        const auto & cache = special ? model->vocab.cache_token_to_piece_special : model->vocab.cache_token_to_piece;
```
Comment (ochafik) [inline llama.cpp:18318]: nit: Maybe we could get away w/ a single cache (built w/ special=true) and early-exit in special case at the top of the function?

```c++

int32_t llama_token_to_piece(const struct llama_model * model, llama_token token, char * buf, int32_t length, bool special) {
    if (!special && llama_is_control_token(model->vocab, token)) {
        return 0;
    }
    // if we have a cache - use it
    if (!model->vocab.cache_token_to_piece.empty()) {
         ....
    }
    ...
```

**Review by ochafik — APPROVED** [review]

**Pull request merged**